flink-runtime/src/main/java/org/apache/flink/runtime/state/DuplicatingCheckpointOutputStream.java

```diff
@@ -0,0 +1,287 @@
+
+	/** Flag if the positional alignment of both streams is checked after each operation. */
+	private static final boolean STRICT_ALIGNMENT_CHECKS = false;
+
+	/** Default buffer size of 8KB. */
+	private static final int DEFAULT_BUFFER_SIZER = 8 * 1024;
+
+	/** Write buffer. */
+	private final byte[] buffer;
+
+	/** Position in the write buffer. */
+	private int bufferIdx;
+
+	/**
+	 * Primary stream for writing the checkpoint data. Failures from this stream are forwarded.
+	 */
+	private final CheckpointStreamFactory.CheckpointStateOutputStream primaryOutputStream;
+
+	/**
+	 * Primary stream for writing the checkpoint data. Failures from this stream are not forwarded until
+	 * {@link #closeAndGetSecondaryHandle()}.
+	 */
+	private final CheckpointStreamFactory.CheckpointStateOutputStream secondaryOutputStream;
+
+	/**
+	 * Stores a potential exception that occurred while interacting with {@link #secondaryOutputStream}
+	 */
+	private IOException secondaryStreamException;
+
+	public DuplicatingCheckpointOutputStream(
+		CheckpointStreamFactory.CheckpointStateOutputStream primaryOutputStream,
+		CheckpointStreamFactory.CheckpointStateOutputStream secondaryOutputStream) throws IOException {
+		this(primaryOutputStream, secondaryOutputStream, DEFAULT_BUFFER_SIZER);
+	}
+
+	public DuplicatingCheckpointOutputStream(
+		CheckpointStreamFactory.CheckpointStateOutputStream primaryOutputStream,
+		CheckpointStreamFactory.CheckpointStateOutputStream secondaryOutputStream,
+		int bufferSize) throws IOException {
+
+		this.primaryOutputStream = Preconditions.checkNotNull(primaryOutputStream);
+		this.secondaryOutputStream = Preconditions.checkNotNull(secondaryOutputStream);
+
+		this.buffer = new byte[bufferSize];
+		this.bufferIdx = 0;
+
+		this.secondaryStreamException = null;
+
+		checkForAlignedStreamPositions();
+	}
+
+	@Override
+	public void write(int b) throws IOException {
+
+		if (buffer.length <= bufferIdx) {
+			flushInternalBuffer();
+		}
+
+		buffer[bufferIdx] = (byte) b;
+		++bufferIdx;
+	}
+
+	@Override
+	public void write(byte[] b) throws IOException {
+
+		write(b, 0, b.length);
+	}
+
+	@Override
+	public void write(byte[] b, int off, int len) throws IOException {
+
+		if (buffer.length <= len) {
+
+			flushInternalBuffer();
+			writeThroughInternal(b, off, len);
+		} else {
+
+			if (buffer.length < len + bufferIdx) {
+				flushInternalBuffer();
+			}
+
+			System.arraycopy(b, off, buffer, bufferIdx, len);
+			bufferIdx += len;
+		}
+	}
+
+	@Override
+	public long getPos() throws IOException {
+		final long referencePos = primaryOutputStream.getPos();
+		return referencePos + bufferIdx;
+	}
+
+	@Override
+	public void flush() throws IOException {
+
+		flushInternalBuffer();
+		primaryOutputStream.flush();
+
+		if (secondaryStreamException == null) {
+			try {
+				secondaryOutputStream.flush();
+			} catch (IOException flushEx) {
+				handleSecondaryStreamOnException(flushEx);
```

Comment:
Why not just catch `throwable` for `secondaryOutputStream` ? Maybe some `RuntimeException` could throw out from it.

Reply:
I would at most catch `Exception`, not the right place to catch for `Throwable`. Will think about it.